packages/lit-dev-server/src/middleware/content-security-policy-middleware.ts

```diff
@@ -0,0 +1,120 @@
+    // into the WASM build, [4] modify or fork the WASM build to implement
+    // string unescaping without eval (needs benchmarking).
+    //
+    // In dev mode, data: scripts are required because @web/dev-server uses them
+    // for automatic reloads.
+    `script-src 'self' 'unsafe-inline' 'unsafe-eval' https://www.googletagmanager.com/gtag/js ${
+      opts.devMode ? ` data:` : ''
+    }`,
+
+    // unpkg.com is needed to allow the Playground worker to fetch dependencies.
+    // TODO(aomarks) After https://crbug.com/1253267 is fixed we can serve a
+    // separate CSP policy just for the worker script (Firefox and Safari
+    // already support this).
+    //
+    // In dev mode, ws: connections are required because @web/dev-server uses
+    // them for automatic reloads.
+    `connect-src 'self' https://unpkg.com/${opts.devMode ? ` ws:` : ''}`,
+
+    // TODO(aomarks) These frame-src directives are only needed for embedding
+    // Playground previews. We can know the exact origin that is being used for
+    // Playground previews, so we could restrict this further by passing that in
+    // as a parameter to the middleware.
+    //
+    // In dev mode, http: is needed for http://localhost Playground iframes.
+    `frame-src https:${opts.devMode ? ' http:' : ''}`,
+
+    // We need 'unsafe-inline' because CodeMirror uses inline styles See
+    // https://discuss.codemirror.net/t/inline-styles-and-content-security-policy/1311/2
+    // TODO(aomarks) Is it possible to know all of the possible inline styles
+    // CodeMirror will use, and provide them here using 'unsafe-hashes'? They
+    // look quite dynamic though.
+    `style-src 'self' 'unsafe-inline'`,
+
+    // TODO(aomarks) We need fonts.gstatic.com for fetching Open Sans (Manrope
+    // is already hosted locally). We should probably just host Open Sans
+    // ourselves, because there is no cross-origin caching benefit (see
+    // https://developers.google.com/web/updates/2020/10/http-cache-partitioning),
+    // and it would be one fewer connection to make.
+    `font-src 'self' https://fonts.gstatic.com/`,
+
+    // TODO(aomarks) We use some data: URLs for SVGs in docs.css. There's
+    // probably a simpler way.
+    // The ytimg.com domain is needed for embedded YouTube videos.
+    `img-src 'self' data: https://i.ytimg.com/`,
+
+    // TODO(aomarks) This could be 'none' if we didn't use <svg><use> elements,
+    // because Firefox does not follow the img-src directive for them, so there
+    // is no other directive to use. See
+    // https://bugzilla.mozilla.org/show_bug.cgi?id=1303364#c4 and
+    // https://github.com/w3c/webappsec-csp/issues/199.
+    `default-src 'self'`,
+
+    ...(opts.reportViolations
+      ? [`report-uri https://csp.withgoogle.com/csp/lit-dev`]
+      : []),
+  ].join('; ');
+
+  return async (ctx, next) => {
+    await next();
+    if (ctx.response.type === 'text/html') {
```

Comment:
Clarification: We only add this header to `text/html` responses?

This was also really helpful to manually test. Was able to see the `content-security-policy-report-only` header being applied on the test domains which is very cool.

Reply:
Yeah, it only needs to be set on top-level HTML responses, since CSP only applies there. (Actually it *should* also apply to worker scripts, which will be helpful for us in the future, but Chrome currently implements this incorrectly [but it is being worked on]). Setting it e.g. on an image would have no effect.